src/sp/it/pl/gui/objects/contextmenu/ImprovedContextMenu.kt

```diff
@@ -103,34 +110,53 @@ class ContextMenuItemSuppliers {
     private val mMany = ClassListMap<ItemsSupply> { fail() }
 
     @Suppress("UNCHECKED_CAST")
-    fun <T: Any> add(type: Class<T>, items: (ImprovedContextMenu<*>, T) -> Sequence<MenuItem>) {
-        mSingle.accumulate(type, items as ItemsSupply)
+    fun <T: Any> add(type: Class<T>, items: ContextMenuBuilder<T>.() -> Unit) {
+        mSingle.accumulate(type) { menu, item ->
+            ContextMenuBuilder(menu, item as T).also { items(it) }.asSequence()
+        }
     }
 
     @Suppress("UNCHECKED_CAST")
-    fun <T: Any> addMany(type: Class<T>, items: (ImprovedContextMenu<*>, Collection<T>) -> Sequence<MenuItem>) {
-        mMany.accumulate(type, items as ItemsSupply)
+    fun <T: Any> addMany(type: Class<T>, items: ContextMenuBuilder<Collection<T>>.() -> Unit) {
+        mMany.accumulate(type) { menu, item ->
+            ContextMenuBuilder(menu, item as Collection<T>).also { items(it) }.asSequence()
+        }
     }
 
-    inline fun <reified T: Any> add(noinline items: (ImprovedContextMenu<*>, T) -> Sequence<MenuItem>) {
+    inline fun <reified T: Any> add(noinline items: ContextMenuBuilder<T>.() -> Unit) {
         add(T::class.java, items)
     }
 
-    inline fun <reified T: Any> addMany(noinline items: (ImprovedContextMenu<*>, Collection<T>) -> Sequence<MenuItem>) {
+    inline fun <reified T: Any> addMany(noinline items: ContextMenuBuilder<Collection<T>>.() -> Unit) {
         addMany(T::class.java, items)
     }
 
     operator fun get(contextMenu: ImprovedContextMenu<*>, value: Any?): Sequence<MenuItem> {
+        logger.trace { "Generating contextmenu for $value" }
         val items1 = mSingle.getElementsOfSuperV(value?.javaClass ?: Void::class.java).asSequence()
-                .map { it(contextMenu, value) }
-                .flatMap { it }
+                .flatMap { it(contextMenu, value) }
         val itemsN = if (value is Collection<*>) {
             mMany.getElementsOfSuperV(value.getElementType()).asSequence()
-                    .map { it(contextMenu, contextMenu.value) }
-                    .flatMap { it }
+                    .flatMap { it(contextMenu, contextMenu.value) }
         } else {
             sequenceOf()
         }
         return items1+itemsN
     }
+
+}
+
+class ContextMenuBuilder<T>(val contextMenu: ImprovedContextMenu<*>, val selected: T): ArrayList<MenuItem>() {
```

Comment:
Composition over inheritance. It seems I will keep saying this until my final days...

This makes ContextMenuBuilder inherit and blatantly expose all of ArrayList API, which is very bad.

Reply:
Well, I first did composition, but why not inheritance? The Builder is just a convenient way of assembling a Collection of Menutems, do it basically is a collection.

Reply:
If you do not find the fact that the builder is now polluted with the ArrayList API disturbing then I do not know what to say. It is also not 'convenient' that the builder has this API. Because it is not its API, its API of something that the builder uses. People are supposed to use the builder, not what the builder is based on.  If you want the builder to have collection-like API then make it yourself (but that would be foolish).

Tomorrow you will realize you need to use a Set and break half the application because the API would change. THis is such an elementary mistake. Do not use inheritance out of convenience! It has disastrous consequences.

I jsut want to really point out that the real problem is not possibility that the API would change (that is merely a demerit of the decision), but the fact that the API that users of the builder will see is not at all what you wanted them to see and the reason is because you were too lazy.

>  it basically is a collection.
No it isnt. You are confusing 'is' and 'has' relationship. The builder uses a collection (and that is not even necessary) and that is mere implementation detail. It is not a collection exactly because users will not use it as a collection nor think of it as such.

Builder having an inner structure and conceptual similarity with collection does not implicate it being a collection. Bottom line, Builder not being compatible (read inherit) with Collection is not disadvantageous. Yes it does limit flexibility, but sometimes that is a good thing! It discards useless freedom the user should not ever have. Suppose everything would be collection, since it is more convenient than calling listOf(). Good luck casting every variable to specific classes in order to disambiguate method calls and tell other developers which method you are actually calling.

It is better to have 'bridge' conversion methods (asSequence(), listOf(), toString(), etc) instead of makign the object be everything. It clearly separates intent by providing only that API which is relevant in given context.